src/sentry/snuba/discover.py

```diff
@@ -69,45 +69,57 @@ def is_real_column(col):
-        project_slug, event_id = reference_event.slug.split(":")
-    except ValueError:
-        raise InvalidSearchQuery("Invalid reference event")
+    columns = [field for field in reference_event.fields if is_real_column(field)]
+    return bulk_find_reference_event([reference_event], columns, "discover.find_reference_event")[0]
 
-    column_names = [resolve_column(col) for col in reference_event.fields if is_real_column(col)]
+
+def bulk_find_reference_event(reference_events, columns, referrer):
     # We don't need to run a query if there are no columns
-    if not column_names:
+    if not columns:
         return None
 
-    try:
-        project = Project.objects.get(
-            slug=project_slug,
-            organization=reference_event.organization,
-            status=ProjectStatus.VISIBLE,
-        )
-    except Project.DoesNotExist:
-        raise InvalidSearchQuery("Invalid reference event")
+    project_ids = set()
+    snuba_filter = eventstore.Filter(event_ids=[])
+    for reference_event in reference_events:
+        try:
+            project_slug, event_id = reference_event.slug.split(":")
+            project = Project.objects.get(
+                slug=project_slug,
+                organization=reference_event.organization,
+                status=ProjectStatus.VISIBLE,
+            )
+        except (ValueError, Project.DoesNotExist):
+            raise InvalidSearchQuery("Invalid reference event")
 
-    start = None
-    end = None
-    if reference_event.start:
-        start = reference_event.start - timedelta(seconds=5)
-    if reference_event.end:
-        end = reference_event.end + timedelta(seconds=5)
+        project_ids.add(project.id)
+        snuba_filter.event_ids.append(event_id)
+
+        start = reference_event.start - timedelta(seconds=5) if reference_event.start else None
+        if start and (snuba_filter.start is None or snuba_filter.start > start):
+            snuba_filter.start = start
+        end = reference_event.end + timedelta(seconds=5) if reference_event.end else None
+        if end and (snuba_filter.end is None or snuba_filter.end > end):
+            snuba_filter.end = end
+
+    snuba_filter.project_ids = list(project_ids)
+    snuba_filter.update_with(resolve_field_list(columns, snuba_filter, auto_fields=False))
+    snuba_filter, _ = resolve_discover_aliases(snuba_filter)
 
     event = raw_query(
-        selected_columns=column_names,
-        filter_keys={"project_id": [project.id], "event_id": [event_id]},
-        start=start,
-        end=end,
+        selected_columns=snuba_filter.selected_columns,
+        filter_keys=snuba_filter.filter_keys,
+        aggregations=snuba_filter.aggregations,
+        start=snuba_filter.start,
+        end=snuba_filter.end,
+        groupby=snuba_filter.groupby,
         dataset=Dataset.Discover,
-        limit=1,
-        referrer="discover.find_reference_event",
+        limit=len(reference_events),
+        referrer=referrer,
     )
```

Comment:
Why can't you just use eventstore to load the event since you know event id and project id ? eventstore is much more efficient for single event loads.

Reply:
Going to be removing these changes around reference event :+1: